firedrake/variational_solver.py

```diff
@@ -305,6 +306,10 @@ def solve(self, bounds=None):
 
         for dbc in problem.dirichlet_bcs():
             dbc.apply(problem.u_restrict)
+            for coeff in coefficients:
+                if isinstance(coeff, Cofunction):
+                    # Apply the DirichletBC to the right hand side of the equation.
+                    dbc.apply(coeff)
```

Comment:
This will have side effects on `Cofunction`s in the problem definition.
```
u = Function(space, name="u")
b = Cofunction(space.dual(), name="b")
...
problem = LinearVariationalProblem(
    inner(trial, test) * dx, b, u, DirichletBC(space, 1, "on_boundary"))
assert b in problem.F.coefficients()  # passes
```